lib/ContentBasedRecommender.js

```diff
@@ -173,57 +204,124 @@ class ContentBasedRecommender {
+   * @private
+   */
+  _calculateSimilaritiesBetweenTwoVectors(documentVectors, targetDocumentVectors, options) {
+    const data = {
+      ...this.initializeDataHash(documentVectors),
+      ...this.initializeDataHash(targetDocumentVectors)
+    }
 
-    // initialize data hash
+    // calculate the similar scores
     for (let i = 0; i < documentVectors.length; i += 1) {
-      const documentVector = documentVectors[i];
-      const { id } = documentVector;
+      if (options.debug) console.log(`Calculating similarity score for document ${i}`);
+
+      for (let j = 0; j < targetDocumentVectors.length; j += 1) {
+        let documentVectorA = documentVectors[i];
+        let targetDocumentVectorB = targetDocumentVectors[j];
+        const idi = documentVectorA.id;
+        const vi = documentVectorA.vector;
+        const idj = targetDocumentVectorB.id;
+        const vj = targetDocumentVectorB.vector;
+        const similarity = vi.getCosineSimilarity(vj);
 
-      data[id] = [];
+        if (similarity > options.minScore) {
+          data[idi].push({
+            ...targetDocumentVectorB,
+            score: similarity
+          });
+          data[idj].push({
+            ...documentVectorA,
+            score: similarity
+          });
+        }
+      }
     }
 
+    this.orderDocuments(data, options);
+
+    return data;
+  }
+
+  initializeDataHash(documentVectors) {
+    return documentVectors.reduce((acc, item) => {
+      acc[item.id] = [];
+      return acc;
+    }, {});
+  }
+
+  _calculateSimilarities(documentVectors, options) {
+    const data = { ...this.initializeDataHash(documentVectors) };
+
     // calculate the similar scores
     for (let i = 0; i < documentVectors.length; i += 1) {
-      if (options.debug) {
-        console.log(`Calculating similarity score for document ${i}`);
-      }
+      if (options.debug) console.log(`Calculating similarity score for document ${i}`);
 
       for (let j = 0; j < i; j += 1) {
-        const idi = documentVectors[i].id;
-        const vi = documentVectors[i].vector;
-        const idj = documentVectors[j].id;
-        const vj = documentVectors[j].vector;
+        let documentVectorA = documentVectors[i];
+        const idi = documentVectorA.id;
+        const vi = documentVectorA.vector;
+        let documentVectorB = documentVectors[j];
+        const idj = documentVectorB.id;
+        const vj = documentVectorB.vector;
         const similarity = vi.getCosineSimilarity(vj);
 
         if (similarity > options.minScore) {
-          data[idi].push({ id: idj, score: similarity });
-          data[idj].push({ id: idi, score: similarity });
+          let docB = {
+            ...documentVectorB,
+            score: similarity
+          };
+
+          delete docB.vector;
+
+          data[idi].push(docB);
+
+          let docA = {
+            ...documentVectorA,
+            score: similarity
+          };
+
+          delete docA.vector;
+
+          data[idj].push(docA);
```

Comment:
it looks it will make the data object very huge as we keep every fields from documentVectorA or documentVectorB